src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -107,10 +151,151 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
+            if (string.Equals("GET", request.Method, StringComparison.OrdinalIgnoreCase))
+            {
+                _responseType = ResponseType.FullReponse;
+            }
+            else if (string.Equals("HEAD", request.Method, StringComparison.OrdinalIgnoreCase))
+            {
+                _responseType = ResponseType.HeadersOnly;
+            }
+            else
+            {
+                return false;
+            }
+
+            // Verify exitence of authorization headers
+            // TODO: The server may indicate that the response to these request are cacheable
+            if (!string.IsNullOrEmpty(request.Headers[HeaderNames.Authorization]))
+            {
+                return false;
+            }
+
+            // Verify request cache-control parameters
+            // TODO: no-cache requests can be retrieved upon validation with origin
+            if (!string.IsNullOrEmpty(request.Headers[HeaderNames.CacheControl]))
+            {
+                if (RequestCacheControl.NoCache || RequestCacheControl.NoStore)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                // Support for legacy HTTP 1.0 cache directive
+                var pragmaHeaderValuese = request.Headers[HeaderNames.Pragma];
+                foreach (var directive in pragmaHeaderValuese)
+                {
+                    if (string.Equals("no-cache", directive, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            // TODO: Verify global middleware settings? Explicit ignore list, range requests, etc.
+            return true;
+        }
+
+        private bool ResponseIsCacheable()
+        {
+            // Only cache pages explicitly marked with public
+            if (!ResponseCacheControl.Public)
+            {
+                return false;
+            }
+
+            // Check no-store
+            if (ResponseCacheControl.NoStore)
+            {
+                return false;
+            }
+
+            // Check no-cache
+            // TODO: Handle no-cache with headers
+            if (ResponseCacheControl.NoCache)
+            {
+                return false;
+            }
+
+            // TODO: public MAY override the cacheability checks for private and status codes
+            
+            // Check private
+            if (ResponseCacheControl.Private)
+            {
+                return false;
+            }
+
+            // Check response code
+            // TODO: RFC also lists 203, 204, 206, 300, 301, 404, 405, 410, 414, and 501 as cacheable by default
+            if (HttpContext.Response.StatusCode != StatusCodes.Status200OK)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private bool ValidateEncoding(ResponseHeaders cachedResponsHeaders)
```

Comment:
This is a little tricky. Technically yes, that's something that you can vary by, but generally this field varies significantly between different requests, even if they can be satisfied by the same encoding. I see in other implementations this field is treated differently so look for an encoding in the cache that satisfies the request. I'm looking into replicating this logic right now.